source/user/lmbench3/doc/lat_fcntl.8
.\" $Id: lat_fcntl.8,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $
.TH LAT_FCNTL 8 "$Date: 2006-11-23 11:25:34 $" "(c)1994-2000 Carl Staelin and Larry McVoy" "LMBENCH"
.SH NAME
lat_fcntl \- fcntl file locking benchmark
.SH SYNOPSIS
.B lat_ctx 
[
.I "-P <parallelism>"
]
[
.I "-W <warmups>"
]
[
.I "-N <repetitions>"
]
.SH DESCRIPTION
.B lat_fcntl
is a client/server program that measures file locking latencies.  The
benchmark alternately locks and unlocks files so that only one of the
client or server is running at a time, similar to ``hot potato''
message passing benchmarks. 
No other work is done in the processes.
.SH ACKNOWLEDGEMENT
Funding for the development of
this tool was provided by Sun Microsystems Computer Corporation.
.SH "SEE ALSO"
lmbench(8), lat_fifo(8), lat_tcp(8), lat_udp(8), lat_unix(8).
.SH "AUTHOR"
Carl Staelin and Larry McVoy
.PP
Comments, suggestions, and bug reports are always welcome.



source/History
Ralink AP SDK - History

Version 4.0.1.0
===============

- [IPv6] Avoid both eth2.2 and eth2 set same IPV6 address after receiving RA packet in gateway mode.
- [IPv6] Fix HNAT misbinding issue (RT6855/6)
- [IPv6] Fix ping fail(ping ::1) issue of loopback interface when we try to enable and disable disable_ipv6 parameter in /proc directory (RT6855/6)
- [Switch] Fix legacy switch(RT3052/RT3352/RT5350) settings vlan 12345 error
- [PPPoE] Kernel panic after PAP fail (RT6855/6)
- [PPTP] Kernel panic after PPTP peer terminated the link (RT6855/6)
- [SPI Flash] There are some noise if we try to program flash while I2S is playing music (RT3052/RT3883/RT3352/RT5350)
- [GPIO] Fix GPIO interrupt handler cannot be trigged correctly (RT6855/6)
- [GUI] Fix WPS bug for test plan
- [NET] Recover irqs_disabled() check under SKB Recycle to fix warning messages, but it decreases RT6855/6's maximum peek throughput
- [Uboot] RT6855/6 does watchdog reset at cold start to fix timer interrupt issue

Version 4.0.0.0
===============

PLATFORM:

- support RT3052/RT3883/RT3352/RT5350 (linux-2.6.21.x)
- support RT6855/RT6856 (linux-2.6.36MT.x)

NEW FUNCTIONS:

- [IPv6 ready logo Phase II] RT6855/RT6856
- [IPv6] support IPv6 Rapid Deployment
- [IPv6] support IPv6 DS-Lite
- [GUI] support two giga phy port
- [GUI] display AP Client site_survey

IMPROVEMENT:

- [IPv6] support ipv6 tftp server
- [PPTP/L2TP] performace tunning at Linux 2.6.21 (PPTP improve 25%, L2TP improve 10% compare to SDK 3.6)
- [SAMBA] significant performance fine tune on full loading scenario (dual-band concurrent AP with USB access)
- [NAND Flash] config page size and address cycle based on chip mode
- [I2S] Add MMAP for user/kernel
- [GUI] support WPS 2.0 

Bug fixed:

- [Kernel NVRAM] memory leakage / initiate failed / free cache / wirte data exceed 1024
- [RAETH] fix STP drop issue on RT3052 chipst ONLY
- [RAETH] fix switch abnormal after ifconfig down&up
- [RAETH] fix PDMA hang issue under heavy load test
- [RAETH] fix TSO bug if USB storage is enabled
- [USB device] system will hang whiling loading gadget driver. Turning off power-saving on both port0 and port1
- [USB device] force host to do re-enumeration to solve "unrecognize device" when re-plug-in in windows platform.
- [IPv6] kernel panic bug if enable ipv6 privacy extension or router prereference
- [IPv6] fix NS packet loss for ipv6 DAD
- [GUI] flush conntrack table while restart/deleting rules
- [NAND Flash] fix nand flash not work for MTD_ANY mode
- [IGMPPROXY] fixed the overflow FATAL bug while handling bit47=1 mac address
- [GPIO] fixed RT3352/RT5350 extended GPIO pins mapping



source/user/lmbench3/src/lat_fcntl.c
#include "bench.h"

/*
 * lat_fcntl.c - file locking test
 *
 * Copyright (c) 1994 Larry McVoy.  Distributed under the FSF GPL with
 * additional restriction that results may published only if
 * (1) the benchmark is unmodified, and
 * (2) the version in the sccsid below is included in the report.
 * Support for this development by Sun Microsystems is gratefully acknowledged.
 */
char	*id = "$Id: lat_fcntl.c,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $\n";

#include "bench.h"

struct	flock lock, unlock;
struct	flock s1, s2;

/*
 * Create two files, use them as a ping pong test.
 * Process A:
 *	lock(1)
 *	unlock(2)
 * Process B:
 *	unlock(1)
 *	lock(2)
 * Initial state:
 *	lock is locked
 *	lock2 is locked
 */

#define	waiton(fd)	fcntl(fd, F_SETLKW, &lock)
#define	release(fd)	fcntl(fd, F_SETLK, &unlock)

struct _state {
	char filename1[2048];
	char filename2[2048];
	int	pid;
	int	fd1;
	int	fd2;
};

void initialize(iter_t iterations, void* cookie);
void benchmark(iter_t iterations, void* cookie);
void cleanup(iter_t iterations, void* cookie);

void
procA(struct _state *state)
{
	if (waiton(state->fd1) == -1) {
		perror("lock of fd1 failed\n");
		cleanup(0, state);
		exit(1);
	}
	if (release(state->fd2) == -1) {
		perror("unlock of fd2 failed\n");
		cleanup(0, state);
		exit(1);
	}
	if (waiton(state->fd2) == -1) {
		perror("lock of fd2 failed\n");
		cleanup(0, state);
		exit(1);
	}
	if (release(state->fd1) == -1) {
		perror("unlock of fd1 failed\n");
		cleanup(0, state);
		exit(1);
	}
}

void
procB(struct _state *state)
{
	if (release(state->fd1) == -1) {
		perror("unlock of fd1 failed\n");
		cleanup(0, state);
		exit(1);
	}
	if (waiton(state->fd2) == -1) {
		perror("lock of fd2 failed\n");
		cleanup(0, state);
		exit(1);
	}
	if (release(state->fd2) == -1) {
		perror("unlock of fd2 failed\n");
		cleanup(0, state);
		exit(1);
	}
	if (waiton(state->fd1) == -1) {
		perror("lock of fd1 failed\n");
		cleanup(0, state);
		exit(1);
	}
}

void 
initialize(iter_t iterations, void* cookie)
{
	char	buf[10000];
	struct _state* state = (struct _state*)cookie;

	if (iterations) return;

	sprintf(state->filename1, "/tmp/lmbench-fcntl%d.1", getpid());
	sprintf(state->filename2, "/tmp/lmbench-fcntl%d.2", getpid());
	state->pid = 0;
	state->fd1 = -1;
	state->fd2 = -1;

	unlink(state->filename1);
	unlink(state->filename2);
	if ((state->fd1 = open(state->filename1, O_CREAT|O_RDWR, 0666)) == -1) {
		perror("create");
		exit(1);
	}
	if ((state->fd2 = open(state->filename2, O_CREAT|O_RDWR, 0666)) == -1) {
		perror("create");
		exit(1);
	}
	unlink(state->filename1);
	unlink(state->filename2);
	write(state->fd1, buf, sizeof(buf));
	write(state->fd2, buf, sizeof(buf));
	lock.l_type = F_WRLCK;
	lock.l_whence = 0;
	lock.l_start = 0;
	lock.l_len = 1;
	unlock = lock;
	unlock.l_type = F_UNLCK;
	if (waiton(state->fd1) == -1) {
		perror("lock1");
		exit(1);
	}
	if (waiton(state->fd2) == -1) {
		perror("lock2");
		exit(1);
	}
	handle_scheduler(benchmp_childid(), 0, 1);
	switch (state->pid = fork()) {
	case -1:
		perror("fork");
		exit(1);
	case 0:
		handle_scheduler(benchmp_childid(), 1, 1);
		for ( ;; ) {
			procB(state);
		}
		exit(0);
	default:
		break;
	}
}

void
benchmark(iter_t iterations, void* cookie)
{
	struct _state* state = (struct _state*)cookie;
	
	while (iterations-- > 0) {
		procA(state);
	}
}

void
cleanup(iter_t iterations, void* cookie)
{
	int i;
	struct _state* state = (struct _state*)cookie;

	if (iterations) return;

	if (state->fd1 >= 0) close(state->fd1);
	if (state->fd2 >= 0) close(state->fd2);
	state->fd1 = -1;
	state->fd2 = -1;

	if (state->pid) {
		kill(state->pid, SIGKILL);
		waitpid(state->pid, NULL, 0);
	}
	state->pid = 0;
}

int
main(int ac, char **av)
{
	int	i;
	int	c;
	int	parallel = 1;
	int	warmup = 0;
	int	repetitions = TRIES;
	struct _state state;
	char *usage = "[-P <parallelism>] [-W <warmup>] [-N <repetitions>]\n";

	/*
	 * If they specified a parallelism level, get it.
	 */
	while (( c = getopt(ac, av, "P:W:N:")) != EOF) {
		switch(c) {
		case 'P':
			parallel = atoi(optarg);
			if (parallel <= 0) lmbench_usage(ac, av, usage);
			break;
		case 'W':
			warmup = atoi(optarg);
			break;
		case 'N':
			repetitions = atoi(optarg);
			break;
		default:
			lmbench_usage(ac, av, usage);
			break;
		}
	}

	state.pid = 0;

	benchmp(initialize, benchmark, cleanup, 0, parallel, 
		warmup, repetitions, &state);
	micro("Fcntl lock latency", 2 * get_n());

	return (0);
}
